### DISCUSSION TITLE
Bi-directional connections not working: I'm probably missing something stupid

### DISCUSSION THREAD

**Discussion by hartogjr** [zeroc-ice/ice#2329] [Q&A] [answered]
L.S.,

The past few months, I've been breaking my head on bi-directional Ice connections. So a "worker" (client) connects to an "orchestrator" (server) that orchestrates workloads to various workers. The worker connects to the orchestrator fine. Then the orchestrator wants to set up the worker (load an algorithm) over the same connection. I keep getting the following server/orchestrator error (with varying UUIDs):
`::Ice::NoEndpointException:
no suitable endpoint available for proxy 'D0454D90-D013-4B5D-B3C8-BE8025B5F766 -t -e 1.1'`
The example code from ZeroC that demonstrates the Ice bi-directional connection works perfectly fine. When I try the same steps in my own code, it keeps giving the above error. It must be something really stupid I'm missing, but I can't seem to figure it out. Help would be greatly appreciated. The details:
* Language: C++ (with compiler clang 17 and Ice C++11 mapping)
* Platform: Gentoo Linux
* Ice version: 3.7.10
* Please see the attached (stripped down to relevant stuff) 
[orchestrator_and_worker.tgz](https://github.com/user-attachments/files/15919302/orchestrator_and_worker.tgz)
for details

**Comment by pepone** [comment]
Hi,

Seems that you forget to provide the `OrchestratorImpl` in your sample code. Are you sure your Orchestartor is using a fixed proxy to callback to the worker, calling `ice_fixed(current.con)` as show in https://github.com/zeroc-ice/ice-demos/blob/0ca97379d3b471fff32c59104c412976ad8b1d05/cpp/Ice/bidir/CallbackI.cpp#L17C35-L17C57

**Reply by hartogjr** [reply]
> Hi,
> 
> Seems that you forget to provide the `OrchestratorImpl` in your sample code. Are you sure your Orchestartor is using a fixed proxy to callback to the worker, calling `ice_fixed(current.con)` as show in https://github.com/zeroc-ice/ice-demos/blob/0ca97379d3b471fff32c59104c412976ad8b1d05/cpp/Ice/bidir/CallbackI.cpp#L17C35-L17C57

Hi Jose,
Thank you for your reply and effort to look in to it. I did already use the `ice_fixed(current.con)`, but not sure I used it correctly. I now included the Impl code (.h and .cpp) for the Orchestrator too. I re-packaged everything again in 
[orchestrator_with_impl_and_worker.tgz](https://github.com/user-attachments/files/15931602/orchestrator_with_impl_and_worker.tgz)

**Reply by pepone** [reply, marked as answer]
The `ice_fixed` method returns a new fixed proxy instead of mutating the existing one. Proxy types in Ice are immutable, and the `ice_xxx` methods return a new proxy with the updated setting.

See https://doc.zeroc.com/ice/3.7/client-side-features/proxies/proxy-methods

In your code the `worker_i->algorithmAsync` is not using the fixed proxy, worker_i is still a regular non-fixed proxy. Then one in the `workers_` map is a fixed proxy.

```
workers_[srcip] = { worker_i->ice_fixed(current_i.con), 0, false, registered, thr };
FI("Registered worker `{:s}' at [{:s}] with {:d} threads",
   Ice::identityToString(worker_i->ice_getIdentity()), srcip, thr);
tasks_.push_back(task(taskType::loadAlg, worker_i->algorithmAsync(alg_), srcip));
```

**Reply by hartogjr** [reply]
This was the solution, thank you very much! Greatly appreciated.
